### PR TITLE
Implement the legacy shared-db relation interface with the ability to scale

### PR DESCRIPTION
**Pull request by shayancanonical**
## Issue
We do not have a functioning mysqlrouter charm.

## Solution
Implement a subordinate mysqlrouter charm with the legacy `shared-db` interface and the ability to scale. The flow of the charm is described below:
- components: consumer application, mysqlrouter, mysql database
- The consumer application requests a database, username and hostname
- mysqlrouter uses the `mysql_client` relation interface (new relation interface) to request a user with `mysqlrouter` extra_roles. this user is used to bootstrap mysqlrouter and create the requested consumer application user.

MySQLRouter is registered and run with systemd for resiliency.

## Context
Below is a dump of a deployment with the legacy mysqlrouter charm
```
(venv) ubuntu@mysql1:~/jhack$ jhack utils show-relation mysql/1:db-router keystone-mysql-router/0:db-router
                                                                                  relation data v0.3                                                                                   
┏━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ relation (id: 4) ┃ keystone-mysql-router                                                           ┃ mysql                                                                          ┃
┡━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│ relation name    │ db-router                                                                       │ db-router                                                                      │
│ interface        │ mysql-router                                                                    │ mysql-router                                                                   │
│ leader unit      │ 0                                                                               │ 1                                                                              │
├──────────────────┼─────────────────────────────────────────────────────────────────────────────────┼────────────────────────────────────────────────────────────────────────────────┤
│ application data │ ╭─────────────────────────────────────────────────────────────────────────────╮ │ ╭────────────────────────────────────────────────────────────────────────────╮ │
│                  │ │ <empty>                                                                     │ │ │ <empty>                                                                    │ │
│                  │ ╰─────────────────────────────────────────────────────────────────────────────╯ │ ╰────────────────────────────────────────────────────────────────────────────╯ │
│ unit data        │ ╭─ keystone-mysql-router/0* ──────────────╮                                     │ ╭─ mysql/1* ──────────────────────────────────────────────────────╮            │
│                  │ │                                         │                                     │ │                                                                 │            │
│                  │ │  MRUP_database         keystone         │                                     │ │  MRUP_allowed_units         "keystone-mysql-router/0"           │            │
│                  │ │  MRUP_hostname         10.154.46.110    │                                     │ │  MRUP_password              "S5zLSGcjYYhrfLpcdszW9nWHVZqHtBWh"  │            │
│                  │ │  MRUP_username         keystone         │                                     │ │  db_host                    "10.154.46.231"                     │            │
│                  │ │  mysqlrouter_hostname  10.154.46.110    │                                     │ │  mysqlrouter_allowed_units  "keystone-mysql-router/0"           │            │
│                  │ │  mysqlrouter_username  mysqlrouteruser  │                                     │ │  mysqlrouter_password       "TTsPxHZhf8HKL6rsTTWm4stMj7bVTdGs"  │            │
│                  │ ╰─────────────────────────────────────────╯                                     │ │  wait_timeout               3600                                │            │
│                  │                                                                                 │ ╰─────────────────────────────────────────────────────────────────╯            │
└──────────────────┴─────────────────────────────────────────────────────────────────────────────────┴────────────────────────────────────────────────────────────────────────────────┘
(venv) ubuntu@mysql1:~/jhack$ jhack utils show-relation keystone-mysql-router/0:shared-db keystone/0:shared-db
                                                                                  relation data v0.3                                                                                   
┏━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ relation (id: 3) ┃ keystone                                                                        ┃ keystone-mysql-router                                                          ┃
┡━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│ relation name    │ shared-db                                                                       │ shared-db                                                                      │
│ interface        │ mysql-shared                                                                    │ mysql-shared                                                                   │
│ leader unit      │ 0                                                                               │ 0                                                                              │
├──────────────────┼─────────────────────────────────────────────────────────────────────────────────┼────────────────────────────────────────────────────────────────────────────────┤
│ application data │ ╭─────────────────────────────────────────────────────────────────────────────╮ │ ╭────────────────────────────────────────────────────────────────────────────╮ │
│                  │ │ <empty>                                                                     │ │ │ <empty>                                                                    │ │
│                  │ ╰─────────────────────────────────────────────────────────────────────────────╯ │ ╰────────────────────────────────────────────────────────────────────────────╯ │
│ unit data        │ ╭─ keystone/0* ─────────────╮                                                   │ ╭─ keystone-mysql-router/0* ────────────────────────╮                          │
│                  │ │                           │                                                   │ │                                                   │                          │
│                  │ │  database  keystone       │                                                   │ │  allowed_units  keystone/0                        │                          │
│                  │ │  hostname  10.154.46.110  │                                                   │ │  db_host        127.0.0.1                         │                          │
│                  │ │  username  keystone       │                                                   │ │  db_port        3306                              │                          │
│                  │ ╰───────────────────────────╯                                                   │ │  password       S5zLSGcjYYhrfLpcdszW9nWHVZqHtBWh  │                          │
│                  │                                                                                 │ │  wait_timeout   3600                              │                          │
│                  │                                                                                 │ ╰───────────────────────────────────────────────────╯                          │
└──────────────────┴─────────────────────────────────────────────────────────────────────────────────┴────────────────────────────────────────────────────────────────────────────────┘
```

## Release Notes
Implement the legacy shared-db relation interface with the ability to scale

## Testing
Integration and unit tests to follow shortly

### CODE REVIEW

**Review by paulomach — APPROVED** [review]
I left just one question, but looks great. The interaction between the relations is not the most easy thing to understand - i would advise us to draw a sequence diagram to visualize the events chain in the contributing.md

**Review by marceloneppel — APPROVED** [review]
LGTM! I have only two questions.

File: templates/mysqlrouter.service.j2 [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+[Unit]
+Description=MySQL Router
+After=network.target
+
+[Service]
+Type=forking
+User=mysql
+Group=mysql
+RuntimeDirectory=mysql
+ExecStart=/var/lib/mysql/{{ charm_app_name }}/start.sh
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply templates/mysqlrouter.service.j2:12]: The `.deb` does indeed not provide the `start.sh` and `stop.sh` scripts. Instead, they're generated when mysqlrouter is bootstrapped. However, mysqlrouter is not registered as a systemd service either, thus the template that enables us to enlist the service. The `charm_app_name` is present just to organize the bootstrap output correctly (`/var/lib/mysql` directory was created during the installation handler)

File: src/relations/shared_db.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+        self.framework.observe(
+            self.charm.on[LEGACY_SHARED_DB].relation_changed, self._on_shared_db_relation_changed
+        )
+        self.framework.observe(
+            self.charm.on[PEER].relation_changed, self._on_peer_relation_changed
+        )
+
+    # =======================
+    #  Handlers
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/relations/shared_db.py]: Addressed in 08554d009b0ef588dc899f8d6c771231ccdca2b5

File: src/relations/database.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Library containing the implementation of the legacy db-router relation."""
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/relations/database.py]: good catch, addressed in 08554d009b0ef588dc899f8d6c771231ccdca2b5

File: src/charm.py [hunk 4]
```diff
@@ -6,87 +6,164 @@
...
-            logger.warning("snapd is not installed. Installing...")
-            self._install_apt_packages(["snapd"])
+    def _on_peer_relation_changed(self, event: RelationChangedEvent) -> None:
+        """Handle the peer relation changed event.
 
-        for package in packages:
+        If a peer is being joined for the first time, bootstrap mysqlrouter
+        and share relevant connection data with the related app.
+        """
+        if isinstance(self.unit.status, WaitingStatus) and self.app_peer_data.get(
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/charm.py:122]: we only want to bootstrap non-leader units once the leader has been bootstrapped (which is what the variable represents). renamed to `MYSQL_ROUTER_LEADER_BOOTSTRAPPED` in 08554d009b0ef588dc899f8d6c771231ccdca2b5

[the remark this answers is not in the document]
Comment (marceloneppel) [reply src/charm.py:122]: Oh, yeah! Thanks for clarifying.

**Comment by shayancanonical** [comment]
Thanks for the reviews! @paulomach Will probably look at better documentation once the new relation has been implemented

**Pull request merged**